File: dotnet/src/DocGen/Clients/PreviewClient.cs
using Dokmatiq.DocGen.Internal;
using Dokmatiq.DocGen.Models;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for PDF page preview rendering.
/// </summary>
public sealed class PreviewClient
{
    private readonly Transport _transport;

    internal PreviewClient(Transport transport) => _transport = transport;

    /// <summary>Render a single page as a PNG image.</summary>
    public byte[] PreviewPage(string filePath, int page = 1, int dpi = 150)
        => _transport.RequestBytes(HttpMethod.Post, "/api/preview/page",
            new { pdfBase64 = FileUtils.ToBase64(filePath), page, dpi });

    /// <summary>Render a single page from bytes.</summary>
    public byte[] PreviewPage(byte[] data, int page = 1, int dpi = 150)
        => _transport.RequestBytes(HttpMethod.Post, "/api/preview/page",
            new { pdfBase64 = FileUtils.ToBase64(data), page, dpi });

    /// <summary>Render multiple pages.</summary>
    public PreviewResponse PreviewPages(string filePath, List<int>? pages = null, int dpi = 150)
    {
        var body = new Dictionary<string, object> { ["pdfBase64"] = FileUtils.ToBase64(filePath), ["dpi"] = dpi };
        if (pages is not null) body["pages"] = pages;
        return _transport.RequestJson<PreviewResponse>(HttpMethod.Post, "/api/preview/pages", body);
    }

    /// <summary>Get total page count.</summary>
    public int PageCount(string filePath)
    {
        var result = _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/preview/page-count", new { pdfBase64 = FileUtils.ToBase64(filePath) });
        return result.TryGetValue("pageCount", out var pc) && pc is System.Text.Json.JsonElement el
            ? el.GetInt32() : 0;
    }
}

File: java/src/main/java/com/dokmatiq/docgen/client/PreviewClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;
import com.dokmatiq.docgen.model.PreviewResponse;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Client for PDF page preview rendering. */
public class PreviewClient {
    private final Transport transport;

    public PreviewClient(Transport transport) {
        this.transport = transport;
    }

    /** Render a single page as a PNG image. */
    public byte[] previewPage(Path file, int page, int dpi) {
        return transport.requestBytes("POST", "/api/preview/page",
                Map.of("pdfBase64", FileUtils.toBase64(file), "page", page, "dpi", dpi));
    }

    /** Render a single page with default DPI (150). */
    public byte[] previewPage(Path file, int page) {
        return previewPage(file, page, 150);
    }

    /** Render a single page (page 1, default DPI). */
    public byte[] previewPage(Path file) {
        return previewPage(file, 1, 150);
    }

    /** Render multiple pages. */
    public PreviewResponse previewPages(Path file, List<Integer> pages, int dpi) {
        var body = new HashMap<String, Object>();
        body.put("pdfBase64", FileUtils.toBase64(file));
        if (pages != null) body.put("pages", pages);
        body.put("dpi", dpi);
        return transport.requestJson("POST", "/api/preview/pages", body, PreviewResponse.class);
    }

    /** Get total page count. */
    public int pageCount(Path file) {
        var result = transport.requestJson("POST", "/api/preview/page-count",
                Map.of("pdfBase64", FileUtils.toBase64(file)), Map.class);
        return ((Number) result.get("pageCount")).intValue();
    }
}
